### PR TITLE
check ssh key pair before use it

### LINKED ISSUE #1103: check ssh key exist before take further actions
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

we allow ssh key as input but we don't check whether it exist or not 
just like Flavor 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/pkg/cloud/services/compute/instance.go#L206

we can add some check before really deploy a VM (createInstance)
**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1103

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

File: pkg/cloud/services/compute/client.go [hunk 1]
```diff
@@ -87,6 +89,15 @@ func (s serviceClient) GetFlavorIDFromName(flavor string) (string, error) {
 	return flavorID, mc.ObserveRequest(err)
 }
 
+func (s serviceClient) GetSSHKeyPair(sshkey string) (*keypairs.KeyPair, error) {
+	mc := metrics.NewMetricPrometheusContext("keypair", "get")
```
Comment (seanschneeweiss) [inline pkg/cloud/services/compute/client.go:93]: Shall we check for empty string in all the "get" functions?

File: pkg/cloud/services/compute/client_mock.go [hunk 2]
```diff
@@ -141,6 +142,20 @@ func (mr *MockClientMockRecorder) GetFlavorIDFromName(arg0 interface{}) *gomock.
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetFlavorIDFromName", reflect.TypeOf((*MockClient)(nil).GetFlavorIDFromName), arg0)
 }
 
+// GetServer mocks base method.
```
Comment (seanschneeweiss) [inline pkg/cloud/services/compute/client_mock.go:145]:
```suggestion
// GetSSHKeyPair mocks base method.
```

File: pkg/cloud/services/compute/instance.go [hunk 3]
```diff
@@ -142,6 +142,11 @@ func (s *Service) createInstanceImpl(eventObject runtime.Object, openStackCluste
 		return nil, fmt.Errorf("error getting flavor id from flavor name %s: %v", instanceSpec.Flavor, err)
 	}
 
+	_, err = s.computeService.GetSSHKeyPair(instanceSpec.SSHKeyName)
+	if err != nil {
+		return nil, fmt.Errorf("error geting ssh key with name %s: %v", instanceSpec.SSHKeyName, err)
```
Comment (seanschneeweiss) [inline pkg/cloud/services/compute/instance.go:147]:
```suggestion
		return nil, fmt.Errorf("error getting SSH key with name %s: %v", instanceSpec.SSHKeyName, err)
```

**Comment by mdbooth** [comment]
What error message does Nova give us on creation if this is not valid? Hopefully it's good enough for user consumption? Can we instead pass that error to the user instead of duplicating this validation?

I'm wary of duplicate validations as they can become an additional source of errors. If we were going to add some additional functionality here, like pausing reconciliation until the SSH key is created, then that would be a different matter. But I think we're just catching an error which OpenStack should already catch and reporting it differently, right?

**Comment by mdbooth** [comment]
> What error message does Nova give us on creation if this is not valid? Hopefully it's good enough for user consumption? Can we instead pass that error to the user instead of duplicating this validation?
> 
> I'm wary of duplicate validations as they can become an additional source of errors. If we were going to add some additional functionality here, like pausing reconciliation until the SSH key is created, then that would be a different matter. But I think we're just catching an error which OpenStack should already catch and reporting it differently, right?

I just tested this and nova says: `Invalid key_name provided.`. How would the UX be if the user saw an error in the machine: `Failed to create OpenStack server: Invalid key_name provided.`?

**Pull request closed without merging**